File: packages/engine/src/modules/sabatier.ts
import { PHYSICAL_CONSTANTS } from "../constants";
import type { SimParams } from "../types";

export interface SabatierOutput {
  secWaterElectrolysis_JPerKg: number;
  grossH2KgPerDay: number;
  h2ConsumedKgPerDay: number;
  h2UnreactedKgPerDay: number;
  o2KgPerDay: number;
  co2ImportedKgPerDay: number;
  ch4KgPerDay: number;
  waterRecycleKgPerDay: number;
  qSabatierW: number;
  kp: number;
}

export function sabatierKp(T: number): number {
  const dH = PHYSICAL_CONSTANTS.dH_sabatier.value;
  const dS = PHYSICAL_CONSTANTS.dS_sabatier.value;
  const R = PHYSICAL_CONSTANTS.R.value;
  return Math.exp(-(dH - T * dS) / (R * T));
}

export function simulateSabatier(params: SimParams, waterKgPerDay: number): SabatierOutput {
  const secWaterElectrolysis_JPerKg =
    (params.Vel * 2 * PHYSICAL_CONSTANTS.F.value) /
    (PHYSICAL_CONSTANTS.M_H2O.value * params.etaFaradayEl);
  const grossH2KgPerDay = waterKgPerDay * (PHYSICAL_CONSTANTS.M_H2.value / PHYSICAL_CONSTANTS.M_H2O.value);
  const o2KgPerDay = waterKgPerDay * ((PHYSICAL_CONSTANTS.M_O2.value / 2) / PHYSICAL_CONSTANTS.M_H2O.value);
  const h2ConsumedKgPerDay = grossH2KgPerDay * params.fConversion;
  const h2UnreactedKgPerDay = grossH2KgPerDay - h2ConsumedKgPerDay;
  const co2ImportedKgPerDay =
    h2ConsumedKgPerDay * (PHYSICAL_CONSTANTS.M_CO2.value / (4 * PHYSICAL_CONSTANTS.M_H2.value));
  const ch4KgPerDay =
    h2ConsumedKgPerDay *
    (PHYSICAL_CONSTANTS.M_CH4.value / (4 * PHYSICAL_CONSTANTS.M_H2.value));
  const waterRecycleKgPerDay =
    h2ConsumedKgPerDay * ((2 * PHYSICAL_CONSTANTS.M_H2O.value) / (4 * PHYSICAL_CONSTANTS.M_H2.value));
  const mdotCH4_molPerS = (ch4KgPerDay / 86400) / PHYSICAL_CONSTANTS.M_CH4.value;
  const qSabatierW = mdotCH4_molPerS * Math.abs(PHYSICAL_CONSTANTS.dH_sabatier.value);
  const kp = sabatierKp(params.Tsabatier);

  return {
    secWaterElectrolysis_JPerKg,
    grossH2KgPerDay,
    h2ConsumedKgPerDay,
    h2UnreactedKgPerDay,
    o2KgPerDay,
    co2ImportedKgPerDay,
    ch4KgPerDay,
    waterRecycleKgPerDay,
    qSabatierW,
    kp
  };
}

File: python/selene_isru/modules/sabatier.py
from __future__ import annotations

import math
from typing import Any

from ..constants import c


def sabatier_kp(temp: float) -> float:
    d_h = c("dH_sabatier")
    d_s = c("dS_sabatier")
    return math.exp(-(d_h - temp * d_s) / (c("R") * temp))


def simulate_sabatier(params: dict[str, Any], water_kg_per_day: float) -> dict[str, float]:
    sec_water_electrolysis_j_per_kg = params["Vel"] * 2 * c("F") / (c("M_H2O") * params["etaFaradayEl"])
    gross_h2_kg_per_day = water_kg_per_day * (c("M_H2") / c("M_H2O"))
    o2_kg_per_day = water_kg_per_day * ((c("M_O2") / 2) / c("M_H2O"))
    h2_consumed_kg_per_day = gross_h2_kg_per_day * params["fConversion"]
    h2_unreacted_kg_per_day = gross_h2_kg_per_day - h2_consumed_kg_per_day
    co2_imported_kg_per_day = h2_consumed_kg_per_day * c("M_CO2") / (4 * c("M_H2"))
    ch4_kg_per_day = h2_consumed_kg_per_day * c("M_CH4") / (4 * c("M_H2"))
    water_recycle_kg_per_day = h2_consumed_kg_per_day * (2 * c("M_H2O")) / (4 * c("M_H2"))
    mdot_ch4_mol_per_s = (ch4_kg_per_day / 86400) / c("M_CH4")
    q_sabatier_w = mdot_ch4_mol_per_s * abs(c("dH_sabatier"))

    return {
        "secWaterElectrolysis_JPerKg": sec_water_electrolysis_j_per_kg,
        "grossH2KgPerDay": gross_h2_kg_per_day,
        "h2ConsumedKgPerDay": h2_consumed_kg_per_day,
        "h2UnreactedKgPerDay": h2_unreacted_kg_per_day,
        "o2KgPerDay": o2_kg_per_day,
        "co2ImportedKgPerDay": co2_imported_kg_per_day,
        "ch4KgPerDay": ch4_kg_per_day,
        "waterRecycleKgPerDay": water_recycle_kg_per_day,
        "qSabatierW": q_sabatier_w,
        "kp": sabatier_kp(params["Tsabatier"]),
    }
